mixnet/config.py

```diff
@@ -2,110 +2,77 @@
-from typing import List, TypeAlias
+from typing import List
 
 from cryptography.hazmat.primitives.asymmetric.x25519 import (
     X25519PrivateKey,
     X25519PublicKey,
 )
-from pysphinx.node import Node
-
-from mixnet.bls import BlsPrivateKey, BlsPublicKey
-from mixnet.fisheryates import FisherYates
+from pysphinx.sphinx import Node as SphinxNode
 
 
 @dataclass
-class MixnetConfig:
-    topology_config: MixnetTopologyConfig
-    mixclient_config: MixClientConfig
-    mixnode_config: MixNodeConfig
-
+class GlobalConfig:
+    """
+    Global parameters used across all nodes in the network
+    """
 
-@dataclass
-class MixnetTopologyConfig:
-    mixnode_candidates: List[MixNodeInfo]
-    size: MixnetTopologySize
-    entropy: bytes
+    membership: MixMembership
+    transmission_rate_per_sec: int  # Global Transmission Rate
+    # TODO: use this to make the size of Sphinx packet constant
+    max_mix_path_length: int
 
 
 @dataclass
-class MixClientConfig:
-    emission_rate_per_min: int  # Poisson rate parameter: lambda
-    redundancy: int
-    topology: MixnetTopology
+class NodeConfig:
+    """
+    Node-specific parameters
+    """
+
+    private_key: X25519PrivateKey
+    mix_path_length: int
+    gossip: GossipConfig
 
 
 @dataclass
-class MixNodeConfig:
-    encryption_private_key: X25519PrivateKey
-    delay_rate_per_min: int  # Poisson rate parameter: mu
+class GossipConfig:
+    # The target number of peers a node should maintain in its p2p network
+    peering_degree: int
 
 
 @dataclass
-class MixnetTopology:
-    # In production, this can be a 1-D array, which is accessible by indexes.
-    # Here, we use a 2-D array for readability.
-    layers: List[List[MixNodeInfo]]
-
-    def __init__(
-        self,
-        config: MixnetTopologyConfig,
-    ) -> None:
-        """
-        Build a new topology deterministically using an entropy and a given set of candidates.
-        """
-        shuffled = FisherYates.shuffle(config.mixnode_candidates, config.entropy)
-        sampled = shuffled[: config.size.num_total_mixnodes()]
+class MixMembership:
+    """
+    A list of public information of nodes in the network.
+    We assume that this list is eventually known to all nodes in the network (e.g. via p2p advertising).
+    """
 
-        layers = []
-        for layer_id in range(config.size.num_layers):
-            start = layer_id * config.size.num_mixnodes_per_layer
-            layer = sampled[start : start + config.size.num_mixnodes_per_layer]
-            layers.append(layer)
-        self.layers = layers
+    nodes: List[NodeInfo]
 
-    def generate_route(self, mix_destination: MixNodeInfo) -> list[MixNodeInfo]:
+    def generate_route(self, num_hops: int, last_mix: NodeInfo) -> list[NodeInfo]:
         """
         Generate a mix route for a Sphinx packet.
         The pre-selected mix_destination is used as a last mix node in the route,
         so that associated packets can be merged together into a original message.
```

Comment:
>  Or does the "merge" refer to a situation when multiple sphinx packets are needed for sending the message?

Yes. This is the implementation from the previous design where a single message is split into multiple Sphinx packets and they are merged in the last mix, if the message is big. In the current design, a single message will fit into a single Sphinx packet because the message will be small. Even if the message is not small enough, we can increase the capacity of Sphinx packet, so that the message is not fragmented. 
However, I didn't remove this fragmentation/merge feature for flexibility. If you think this makes confusion and make our design complicated, I can remove this and make sure that a message is transformed to a single Sphinx packet.

Reply:
2d703ca I just removed all fragment-related code. In the new NomMix spec, we're assuming that a single message is small enough to fit into a single Sphinx packet. We're designing bandwidth-related properties based on this assumption. So, it's more reasonable to not have fragment-related code in this executable spec. If the message needs to be bigger, we can modify the Sphinx spec to accept a bigger message for a single Sphinx packet, as I mentioned [above](https://github.com/logos-co/nomos-specs/pull/98#discussion_r1673247324).

Reply:
Great, thank you! I understand the flexibility of the fragmentation, however, as you also note, we would need to gossip every fragment of the message. This is not a big issue but could significantly increase the message dissemination as each fragment will be independently delayed on every node, and the receiving node will need to wait for all fragments to proceed (with cryptographical mix).